### PR TITLE
Remove help verbose 8497

### PR DESCRIPTION
**Pull request by vinod-tahelyani**
Closes: #8497 

### **The new output looks something like-**

**Commands**
* You can run commands with "serverless" or the shortcut "sls"
* Pass "--no-color" to disable CLI colors
* Pass "--help" after any <command> for contextual help

**Interactive Quickstart**
* Run serverless (or shortcut sls) without any arguments to initialize an interactive setup
  of functionalities related to given service or current environment
* Pass "--help-interactive" for contextual help on interactive CLI options

**Serverless Components**
* Run serverless (or shortcut sls) in context of a component service to initialize a components CLI
* Pass "--help-components" for contextual help on Serverless Components

**Framework**
* Documentation: http://slss.io/docs

**Environment Variables**
* Set SLS_DEBUG=* to see debugging logs
* Set SLS_WARNING_DISABLE=* to hide warnings from the output
* Set SLS_MAX_CONCURRENT_ARTIFACTS_UPLOADS to control the maximum S3 upload SDK requests that are sent in parallel during the deployment of the service's artifacts. The default is 3. Note: increasing this too high might, actually, downgrade the overall upload speed

**General Commands**

config ........................ Configure Serverless
config credentials ............ Configures a new provider profile for the Serverless Framework
config tabcompletion install .. Install a <tab> completion for chosen shell
config tabcompletion uninstall  Uninstall a <tab> completion for chosen shell
create ........................ Create new Serverless service
dashboard ..................... Open the Serverless dashboard
deploy ........................ Deploy a Serverless service
deploy function ............... Deploy a single function from the service
deploy list ................... List deployed version of your Serverless Service
deploy list functions ......... List all the deployed functions and their versions
generate-event ................ Generate event
info .......................... Display information about the service
install ....................... Install a Serverless service from GitHub or a plugin from the Serverless registry
invoke ........................ Invoke a deployed function
invoke local .................. Invoke function locally
login ......................... Login or sign up for Serverless
logout ........................ Logout from Serverless
logs .......................... Output the logs of a deployed function
metrics ....................... Show metrics for a specific function
output ........................
output get .................... Get value of dashboard deployment profile parameter
output list ................... List all dashboard deployment profile parameters
package ....................... Packages a Serverless service
param .........................
param get ..................... Get value of dashboard service output
param list .................... List all dashboard service outputs
plugin ........................ Plugin management for Serverless
plugin install ................ Install and add a plugin to your service
plugin uninstall .............. Uninstall and remove a plugin from your service
plugin list ................... Lists all available plugins
plugin search ................. Search for plugins
print ......................... Print your compiled and resolved config file
remove ........................ Remove Serverless service and all resources
rollback ...................... Rollback the Serverless service to a specific deployment
rollback function ............. Rollback the function to a specific version
ses-template .................. Manage AWS SES templates
ses-template deploy ........... Sync email templates to AWS SES
ses-template delete ........... Delete email template from AWS SES by given name
ses-template list ............. List email templates from AWS SES
slstats ....................... Enable or disable stats
studio ........................ Develop a Serverless application in the cloud.
test .......................... Run HTTP tests

**Plugins**
DynamoDBAutoscaling, ServerlessSesTemplate

**Commands by plugin**

ServerlessSesTemplate
  ses-template .................. Manage AWS SES templates
  ses-template deploy ........... Sync email templates to AWS SES
  ses-template delete ........... Delete email template from AWS SES by given name
  ses-template list ............. List email templates from AWS SES

sls-help-1
sls-help-2
sls-help-3

### CODE REVIEW

**Comment by vinod-tahelyani** [comment]
@medikoo pls review this

File: lib/classes/PluginManager.js [hunk 1]
```diff
@@ -90,8 +90,10 @@ class PluginManager {
     this.cliCommands = commands;
   }
 
-  addPlugin(Plugin) {
+  addPlugin(Plugin, isExternal = false) {
     const pluginInstance = new Plugin(this.serverless, this.cliOptions);
+    // isExternal differentiates plugin as OOTB or not
+    Object.assign(pluginInstance, { isExternal });
```
Comment (medikoo) [inline lib/classes/PluginManager.js:96]: It'll be more transparent to create `this.externalPlugins` collection (as set instance), and just group all the plugins there.

Assigning a property to a plugin instance, while should in most cases be harmless, it's destructive, note that we're interferring into plugin API, we can't be sure if we won't break things

File: lib/classes/CLI.js [hunk 2]
```diff
@@ -242,14 +247,66 @@ class CLI {
...
-      );
-      this.consoleLog(sortedPlugins.map(plugin => plugin.constructor.name).join(', '));
+    if (externalSortedPlugins.length) {
+      // print all the installed plugins
+      this.consoleLog(chalk.yellow.underline('Plugins'));
+
+      this.consoleLog(externalSortedPlugins.map(plugin => plugin.constructor.name).join(', '));
+
+      this.consoleLog('');
+      this.consoleLog(chalk.yellow.underline('Commands by plugin'));
```
Comment (medikoo) [inline lib/classes/CLI.js:261]: Let's show this section, only if we have any commands added by plugins

File: lib/classes/CLI.js [hunk 3]
```diff
@@ -242,14 +247,66 @@ class CLI {
...
+      Object.entries(pluginCommands).forEach(([plugin, details]) => {
+        this.consoleLog(plugin);
+        details.forEach(cmd => {
+          // display command usage with single(1) indent
+          this.displayCommandUsage(cmd, cmd.key.split(':').join(' '), 1);
+        });
+        this.consoleLog('');
+      });
     } else {
       this.consoleLog('No plugins added yet');
```
Comment (medikoo) [inline lib/classes/CLI.js:311]: I assume this is to be displayed, when no external plugins are configured (?)

I think we can skip this message (In current implementation it was never shown, as there were always some internal plugins)

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @vinod-tahelyani ! It looks very good. I've proposed few improvements. Let me know what you think

File: lib/classes/CLI.js [hunk 4]
```diff
@@ -231,9 +228,17 @@ class CLI {
...
+    this.consoleLog(chalk.yellow.underline('General Commands'));
     this.consoleLog('');
     if (Object.keys(this.loadedCommands).length) {
-      Object.entries(this.loadedCommands).forEach(([command, details]) => {
+      const internalCommands = Object.entries(this.loadedCommands).filter(
+        command => command && !command.isExternal
+      );
+      const sortedInternalCommands = internalCommands.sort((command1, command2) =>
+        command1[0].localeCompare(command2[0])
+      );
```
[the remark this answers is not in the document]
Comment (vinod-tahelyani) [reply lib/classes/CLI.js:240]: The current help output doesn't seems to have sorted list

> config ........................ Configure Serverless
config credentials ............ Configures a new provider profile for the Serverless Framework
config tabcompletion install .. Install a <tab> completion for chosen shell
config tabcompletion uninstall  Uninstall a <tab> completion for chosen shell
create ........................ Create new Serverless service
install ....................... Install a Serverless service from GitHub or a plugin from the Serverless registry
package ....................... Packages a Serverless service
deploy ........................ Deploy a Serverless service
deploy function ............... Deploy a single function from the service
deploy list ................... List deployed version of your Serverless Service
deploy list functions ......... List all the deployed functions and their versions
invoke ........................ Invoke a deployed function

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/classes/CLI.js:240]: Ok, thanks for clarification, indeed it'll be an improvement

File: lib/classes/PluginManager.js (hunk 1 above)
Comment (vinod-tahelyani) [reply lib/classes/PluginManager.js:96]: u mean keeping `this.plugins `as it is and adding `this.externalPlugins`, right ?

Comment (medikoo) [reply lib/classes/PluginManager.js:96]: Yes, exactly

File: lib/classes/CLI.js (hunk 3 above)
Comment (vinod-tahelyani) [reply lib/classes/CLI.js:311]: but in this case under 'Plugin', we do not list internal plugins, So this can be seen right ?

Comment (medikoo) [reply lib/classes/CLI.js:311]: Yes, and we don't want to be seen. Message is weird "No plugins added yet" (why yet?).

I think best way is to not show "Plugins" section at all

**Comment by vinod-tahelyani** [comment]
@medikoo done with all the changes, pls tell me what do u think

File: lib/classes/CLI.js [hunk 5]
```diff
@@ -16,13 +16,18 @@ class CLI {
     this.serverless = serverless;
     this.inputArray = inputArray || null;
     this.loadedPlugins = [];
+    this.loadedExternalPlugins = new Set();
```
Comment (medikoo) [inline lib/classes/CLI.js:19]: Let's not introduce this collection (we can simply access the external plugins from plugin manager via `this.serverless.pluginManager.externalPlugins`)

I understand that you've tried to maintain the existing convention, however some of the conventions as found in the code-base are old, unnecessarily verbose (imply maintenance cost, which could be avoided), and we try to not follow them with new code.

File: lib/classes/CLI.js [hunk 6]
```diff
@@ -16,13 +16,18 @@ class CLI {
...
     this.loadedCommands = {};
   }
 
   setLoadedPlugins(plugins) {
     this.loadedPlugins = plugins;
   }
 
+  setLoadedExternalPlugins(externalPlugins) {
```
Comment (medikoo) [inline lib/classes/CLI.js:27]: Let's not define this method

File: lib/classes/CLI.js [hunk 7]
```diff
@@ -231,9 +233,17 @@ class CLI {
       )
     );
 
+    this.consoleLog('');
+    this.consoleLog(chalk.yellow.underline('General Commands'));
     this.consoleLog('');
     if (Object.keys(this.loadedCommands).length) {
-      Object.entries(this.loadedCommands).forEach(([command, details]) => {
+      const internalCommands = Object.entries(this.loadedCommands).filter(
+        command => command && !command.isExternal
```
Comment (medikoo) [inline lib/classes/CLI.js:241]: I think this doesn't filter as expected, as in `Object.entries`  result, each item is a `[key, value]` array, so `command` here is an array.

Simple fix would be to use `Object.values` instead, which is also more appriopriate in this case

File: lib/classes/CLI.js [hunk 8]
```diff
@@ -231,9 +233,17 @@ class CLI {
       )
     );
 
+    this.consoleLog('');
+    this.consoleLog(chalk.yellow.underline('General Commands'));
     this.consoleLog('');
     if (Object.keys(this.loadedCommands).length) {
```
Comment (medikoo) [inline lib/classes/CLI.js:239]: Let's remove this `if` clause (in all cases there'll be some commands)

File: lib/classes/CLI.js [hunk 9]
```diff
@@ -242,16 +252,68 @@ class CLI {
 
     this.consoleLog('');
 
-    // print all the installed plugins
-    this.consoleLog(chalk.yellow.underline('Plugins'));
+    const externalSortedPlugins = this.loadedPlugins
```
Comment (medikoo) [inline lib/classes/CLI.js:255]: We can just name it `externalPlugins` (`externalSortedPlugins` would make sense if we would also host _unsorted_ collection, but here it's not the case)

File: lib/classes/PluginManager.js [hunk 10]
```diff
@@ -114,21 +119,27 @@ class PluginManager {
...
+    this.loadCommands(pluginInstance, options);
     this.loadHooks(pluginInstance);
     this.loadVariableResolvers(pluginInstance);
 
     this.plugins.push(pluginInstance);
   }
 
   loadAllPlugins(servicePlugins) {
+    const internalPluginOptions = { isExternal: false };
+    const externalPluginOptions = { isExternal: true };
```
Comment (medikoo) [inline lib/classes/PluginManager.js:131]: Let's not create this variables, we can simply hardcode `{ isExternal: true }` as second argument when adding external plugins and in other cases do not pass any _options_ at all (as _options_ are _optional_)

File: lib/classes/PluginManager.js [hunk 11]
```diff
@@ -259,7 +270,7 @@ class PluginManager {
     return Object.assign({}, details, { key, pluginName, commands });
   }
 
-  loadCommands(pluginInstance) {
+  loadCommands(pluginInstance, options) {
```
Comment (medikoo) [inline lib/classes/PluginManager.js:273]: Let's mark `options` as optional argument, by assigning a _default_ value

File: lib/classes/PluginManager.js [hunk 12]
```diff
@@ -275,7 +286,7 @@ class PluginManager {
           }
           return event;
         });
-        this.commands[key] = _.merge({}, this.commands[key], command);
+        this.commands[key] = _.merge({}, this.commands[key], command, options);
```
Comment (medikoo) [inline lib/classes/PluginManager.js:289]: It'll be more future proof, to specifically pass `{ isExternal: options.isExternal }` (by design `options` are `loadCommands` operation options, and not command extension, and here we treat it as if it's command extension)

File: lib/classes/PluginManager.js [hunk 13]
```diff
@@ -484,6 +495,10 @@ class PluginManager {
     return this.plugins;
   }
 
+  getExternalPlugins() {
```
Comment (medikoo) [inline lib/classes/PluginManager.js:498]: No need for that method

**Review by medikoo — CHANGES_REQUESTED** [review]
Great thanks @vinod-tahelyani! It looks way better, we're close, Please see my final suggestions

**Comment by vinod-tahelyani** [comment]
@medikoo done

**Review by medikoo — APPROVED** [review]
Thank you @vinod-tahelyani ! That looks great

**Pull request merged**